MaterialLib/PorousMedium/Permeability/Permeability.h

```diff
@@ -0,0 +1,103 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2017, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ * \file:   Permeability.h
+ *
+ */
+
+#pragma once
+
+#include <Eigen/Dense>
+
+#include "BaseLib/Error.h"
+#include "ProcessLib/Parameter/Parameter.h"
+
+namespace MaterialLib
+{
+namespace PorousMedium
+{
+/// The class implements a basic permeability model that employs a parameter
+/// (for instance a constant parameter or mesh cell dependend parameter) to fill
+/// the intrinsic permeability tensor.
+class Permeability
+{
+public:
+    explicit Permeability(
+        ProcessLib::Parameter<double> const& permeability_parameter,
+        int const dimension)
+        : _permeability_parameter(permeability_parameter), _dimension(dimension)
+    {
+        auto const number_of_components =
+            _permeability_parameter.getNumberOfComponents();
+
```

Comment:
How about to use the following lines to replace the long switch case?
```
if (number_of_components != (_dimension *_dimension))
{
     OGS_FATAL(
                "The given parameter has %d component, but the "
                "permeability tensor is defined for a %d dimensional "
                "problem.",
                        _dimension * _dimension, _dimension);
}
```

Reply:
Both the current code and your suggested code handle only tensors that are dense, i.e., 1d - 1 component, 2d - 4 components and 3d - 9 components. I wrote this with the idea in mind to generalize this also to diagonal and symmetric tensor cases, i.e., 2d tensor that can have 2 diagonal components or 3 components in the symmetric case and 3d tensor with 3 or 6 components. I'll try to change the code here and in createPermeabilityModel.cpp. What do you think, should I do it now or after your PR #1970 is rebased, reviewed and merged?

Reply:
I see. 
Maybe the diagonal elements are only needed for input together with an optional input of system rotational angles for the anisotropic permeability in local system, e.g. permeability in a inclined bedding plane.  In the practice, there is seldom case that all elements of a permeability tensor are given or measured directly. The input of diagonal elements with an optional input of system rotational angles  sounds more practical for permeability, thermal conductivity or concentration diffusion tensor. We may need a general input and calculation (for possible system rotation)  interface for such tensors. So far, I am satisfied with your current implementation.